File: Valid Parentheses/Valid.py
# Solution.isValid()
# https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        # Создаю словарь для проверки (сходства) ключа и значения для разных скобок
        left_check_right = {'[': ']', '(': ')', '{': '}'}
        # Создаю лист, куда буду помещать скобки и сравнивать 
        arr = []
        # Создаю цикл для перебора элементов 
        for c in s:
            # Если элемент находится в словаре, то добавляю элемент в лист
            if c in left_check_right:
                arr.append(left_check_right[c])
            # Если лист равен нулю или не равен элементу, возвратить false 
            elif len(arr) == 0 or arr.pop() != c:
                return False
        return len(arr) == 0

File: Valid Parentheses/Valid.js
// Почти такая же логика как на питоне
const isValid = (s) => {
  const arr = [];
  const map = new Map([
    ["(", ")"],
    ["[", "]"],
    ["{", "}"],
  ]);

  for (let i = 0; i < s.length; i += 1) {
    // Has по сути тоже самое что и includes
    if (map.has(s[i])) {
      arr.push(map.get(s[i]));
    } else if (s[i] !== arr.pop()) {
      return false;
    }
  }

  return arr.length === 0;
};
